engines/private/private.cpp

```diff
@@ -737,6 +750,158 @@ void PrivateEngine::selectMask(Common::Point mousePos) {
+		if (sym->u.val) {
+			if (inMask(_locationMasks[i].surf, mousePos)) {
+				for (uint j = 0; j < _diaryPages.size(); j++) {
+					if (_diaryPages[j].locationID == totalLocations + 1) {
+						_currentDiaryPage = j;
+						break;
+					}
+				}
+	
+				_numberClicks++;
+				_nextSetting = _locationMasks[i].nextSetting;
+
+				return true;
+			}
+			i++;
+		}
+		totalLocations++;
+	}
+
+	return false;
+}
+
+bool PrivateEngine::selectDiaryNextPage(Common::Point mousePos) {
+	mousePos = mousePos - _origin;
+	if (mousePos.x < 0 || mousePos.y < 0)
+		return false;
+
+	if (_diaryNextPageExit.rect.contains(mousePos)) {
+		_currentDiaryPage++;
+		_nextSetting = _diaryNextPageExit.nextSetting;
+
+		return true;
+	}
+
+	return false;
+}
+
+bool PrivateEngine::selectDiaryPrevPage(Common::Point mousePos) {
+	mousePos = mousePos - _origin;
+	if (mousePos.x < 0 || mousePos.y < 0)
+		return false;
+
+	if (_diaryPrevPageExit.rect.contains(mousePos)) {
+		_currentDiaryPage--;
+		_nextSetting = _diaryPrevPageExit.nextSetting;
+
+		return true;
+	}
+
+	return false;
+}
+
+bool PrivateEngine::selectMemory(Common::Point mousePos) {
+	if (_memoryMasks.size() == 0) { 
+		return false;
+	}
+
+	for (uint i = 0; i < _memoryMasks.size(); i++) {
+		if (inMask(_memoryMasks[i].surf, mousePos)) {
+			_masks.clear();
+			_memoryMasks.clear();
+			_exits.clear();
+			_nextMovie = _diaryPages[_currentDiaryPage].memories[i].movie;
+			_nextSetting = "kDiaryMiddle";
+			loadImage("inface\\general\\inface2.bmp", 0, 0);
+			return true;
+		}
+	}
+
+	return false;
+}
+
+void PrivateEngine::addMemory(Common::String path) {
+	size_t index = path.findLastOf('\\');
+	Common::String location = path.substr(index + 2, 2);
+
+	Common::String imagePath;
+
+	// Paths to the global folder have a different pattern from other paths
+	if (path.contains("global")) {
+		if (path.contains("spoc00xs")) {
+			imagePath = "inface\\diary\\ss_icons\\global\\transiti\\ipoc00.bmp";
+		} else {
+			imagePath = "inface\\diary\\ss_icons\\global\\transiti\\animatio\\mo\\imo" + path.substr(index + 4, 3) + ".bmp";
+		}
+	} else {
+		// First letter after the last \ is an s, which isn't needed; next 2 are location; and the next 3 are what image to use
+		imagePath = "inface\\diary\\ss_icons\\" + location + "\\i" + location + path.substr(index + 4, 3) + ".bmp";
+	}
+
+
+	if (!Common::File::exists(convertPath(imagePath))) {
+		return;
+	}
+
+	MemoryInfo memory;
+	memory.movie = path;
+	memory.image = imagePath;
+
+	for (int i = _diaryPages.size() - 1; i >= 0; i--) {
+		if (_diaryPages[i].locationName == location) {
+			if (_diaryPages[i].memories.size() == 6) {
+				DiaryPage diaryPage = DiaryPage();
+				diaryPage.locationName = location;
+				diaryPage.locationID = _diaryPages[i].locationID;
+				diaryPage.memories.push_back(memory);
+				_diaryPages.insert_at(i + 1, diaryPage);
+				return;
+			}
+			_diaryPages[i].memories.push_back(memory);
+
+			return;
+		}
+	}
+
+	DiaryPage diaryPage = DiaryPage();
```

Comment:
```suggestion
	DiaryPage diaryPage;
```
Avoids unnecessary object copying